blocks/ai-chat/src/complete-chat.tsx

```diff
@@ -0,0 +1,486 @@
+  aiChatBlockEntity: AIChatBlock;
+  initialRootRequestId?: RequestId;
+  initialCompleteChatRequests?: CompleteChatRequest[];
+  initialCompleteChatResponses?: CompleteChatResponse[];
+}> = ({
+  readonly,
+  aiChatBlockEntity,
+  initialRootRequestId,
+  initialCompleteChatRequests,
+  initialCompleteChatResponses,
+}) => {
+  const blockRootRef = useRef<HTMLDivElement>(null);
+
+  const {
+    metadata: {
+      recordId: { entityId: aiChatBlockEntityId },
+    },
+  } = aiChatBlockEntity;
+
+  const { serviceModule } = useServiceBlockModule(blockRootRef);
+  const { graphModule } = useGraphBlockModule(blockRootRef);
+
+  const [chatModel, setChatModel] = useState<ChatModelId>(defaultChatModelId);
+
+  const aiChatBlockEntityPresetSystemPromptId = useMemo<
+    SystemPromptId | undefined
+  >(() => {
+    const value =
+      aiChatBlockEntity.properties[propertyTypeBaseUrls.presetSystemPromptId];
+
+    return value && isSystemPromptId(value) ? value : undefined;
+  }, [aiChatBlockEntity]);
+
+  const aiChatBlockEntityChatAiModel = useMemo<ChatModelId | undefined>(() => {
+    const value =
+      aiChatBlockEntity.properties[propertyTypeBaseUrls.openAIChatModelName];
+
+    return value && isChatModelId(value) ? value : undefined;
+  }, [aiChatBlockEntity]);
+
+  const [systemPromptId, setSystemPromptId] = useState<SystemPromptId>(
+    aiChatBlockEntityPresetSystemPromptId ?? defaultSystemPromptId,
+  );
+
+  const systemPrompt = useMemo<OpenAIChatMessage<"system">>(
+    () => ({ role: "system", content: systemPrompts[systemPromptId] }),
+    [systemPromptId],
+  );
+
+  /**
+   * @todo: consider unifying the `completeChatRequests` and `completeChatResponses`
+   * arrays into a single `messages` array to simplify the logic in this file. Potentially
+   * `RequestId` and `ResponseId` could be unified in a single `MessageId` as part of this.
+   */
+
+  const [completeChatRequests, setCompleteChatRequests] = useState<
+    CompleteChatRequest[]
+  >(initialCompleteChatRequests ?? []);
+
+  const [completeChatResponses, setCompleteChatResponses] = useState<
+    CompleteChatResponse[]
+  >(initialCompleteChatResponses ?? []);
+
+  const [rootChatRequestId, setRootChatRequestId] = useState<
+    RequestId | undefined
+  >(initialRootRequestId);
+
+  const messageThread = useMemo(() => {
+    const rootChatRequest = completeChatRequests.find(
+      ({ id }) => id === rootChatRequestId,
+    );
+
+    return rootChatRequest
+      ? constructMessageThread({
+          currentRequest: rootChatRequest,
+          allRequests: completeChatRequests,
+          allResponses: completeChatResponses,
+        })
+      : [];
+  }, [rootChatRequestId, completeChatRequests, completeChatResponses]);
+
+  const [loading, setLoading] = useState<boolean>(false);
+
+  const updateAiChatBlockEntity = useCallback(
+    async (params: {
+      updatedProperties: Partial<AIChatBlock["properties"]>;
+    }) => {
+      const { updatedProperties } = params;
```

Comment:
Shouldn't the request fail anyway because the EA is enforcing whether or not they can be called? (cc @CiaranMn)

If this is inconsistent with a previous decision I'm happy to change this, but I feel like the block should be ensuring that an update method can't get called by a user in the UI in the first place, rather than abort in the function when it has gotten called. This could also help us catch when this is happening more easily, as an error should be thrown by the EA.

I also don't think it's the block's responsibility to enforce such a safety measure, it's the responsibility of the EA. But if my understanding is incorrect please let me know.